### DISCUSSION TITLE
How to sinchronize transactions with sorted lists?

### DISCUSSION THREAD

**Discussion by faermanj** [infinispan/infinispan#12047] [Q&A]
Hello,

I'm trying to build a simple transaction processing demo (credit / debit / balance) using infinispan (embedded) in two API nodes.
I've setup pessimistic locking and timeout, like this:
https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl/blob/main/rinhadebackend/src/main/archived/ispn/CacheProducer.java

And trying to process transactions like this...

write transaction (credit/debit):
https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl/blob/69753ac034402d7463b55aa5e18c77af7ca65c76/rinhadebackend/src/main/archived/ispn/CachedServlet.java#L115

read transaction (balance, 10 latest)
https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl/blob/69753ac034402d7463b55aa5e18c77af7ca65c76/rinhadebackend/src/main/archived/ispn/CachedServlet.java#L227

However, the application fails to properly synchronize access to the data, resulting in overrides and failures to keep balance consistent.

There are no errors or exceptions on the consoles, all seems to be working as expected.

I'm quite new to infinispan and not sure how to proceed, any clues would be greatly appreciated, because i have none :)

Thanks in advance,
Julio

**Comment by pruivo** [comment]
It overrides because you are modifying the same instance. Infinispan does not do any cloning when storing or loading the instance. My advice is to make your  `Cliente` class immutable or make sure you create a new instance before modifying it. 

To be on the safer side, both keys and values should be immutable.

**Reply by faermanj** [reply]
Thanks, i'll try that.

**Reply by faermanj** [reply]
I changed the transaction method to create a new instance of the object with customer data and clone the list of transactions.

Like this:
https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl/blob/c7072a9474ffbe977d7446d3c101544e60648f76/rinhadebackend/src/main/java/caravanacloud/ispn/Cliente.java#L79

However, it seems that the list of transactions is being changed in both servers and being overwritten.

Is there a way to have infinispan wait until the other server releases the lock, or a better design for keeping sorted lists?

**Reply by pruivo** [reply]
I can't spot anything wrong with your code. Can you try to reproduce the issue with trace logging for the `org.infinispan` package?

**Reply by faermanj** [reply]
The issue is that, if I send several transaction requests simultaneously, the balance gets corrupted. 
I believe that would be because the lists are obtained, changed, and put back in overlap.
From the application, what I can see is the stress tests failing (they pass on relational databases) with errors indicating that the sum and order of transactions is different from the expected.

**Reply by faermanj** [reply]
An example for reference. No errors in server or infinispan logs.
![image](https://github.com/infinispan/infinispan/assets/356476/2d6c11bb-b61f-4fc9-aac5-54bcd11e630e)

**Reply by pruivo** [reply]
If you provide the logs, I can figure out what's going on.

**Reply by faermanj** [reply]
@pruivo cool, i appreciate the helpout, i'll record a brief demo

**Reply by faermanj** [reply]
Here's a shot of the console logs, it's stable so i believe infinispan is working as expected, but my code or configuration is missing something.
![image](https://github.com/infinispan/infinispan/assets/356476/f187cca9-19cb-4954-98ed-ec9647cd457e)

**Comment by faermanj** [comment]
Here's a brief demonstration of the issue:
https://www.loom.com/share/faf63e1e1464409abd2b07d27105551c?sid=5a3ead4f-0551-474e-9084-2b0210dce71c

And here's the code:
https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl/blob/main/rinhadebackend-ispn/src/main/java/caravanacloud/ispn/CachedServlet.java

Thanks!

**Reply by pruivo** [reply]
May I ask about https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl/blob/a18a20da050d909ad12e8702cdf7b67a4c7209ca/rinhadebackend-ispn/src/main/java/caravanacloud/ispn/Cliente.java#L93-L95 ?
You're blindly dropping transactions and couldn't that explain why the balance doesn't add up in your stress test?

**Reply by faermanj** [reply]
Sure! The balance request (extrato) is expected to return the current balance and the 10 latest transactions, that's why the list is cut to this size and created with the comparator by date.

**Reply by pruivo** [reply]
Is there an easier way for me to try and debug it? 
I have a couple of concerts:
* The PriorityQueue is sorted from the most recent to the older transaction but, the `poll()` method will remove the most recent transaction.
* The project says, the JSON response must have the latest 10 transactions, but you need all of them to compute the correct balance.

ps. the challenge looks interesting :+1:

**Reply by faermanj** [reply]
Sure thing, I'll post a new recording of how to run the working and non-working versions, it's all super simple and ready to run on gitpod.
The balance aspect is correct, similar to a real bank, I can see my all-time balance and latest transactions, expected not to match, that's fine. What I'm suspecting is that, somehow, the list is being changed simultaneously on both servers.
Thanks a lot for the helpout!

**Comment by faermanj** [comment]
Here's how to run the project, how I expected transactions to work and how it's failing.
Thanks in advance for the helpout!
https://www.loom.com/share/6a78f8904caf429d850af786fcd84463

**Reply by faermanj** [reply]
Here's a link to the repo:
https://github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl

And to start it on a gitpod:
https://gitpod.io/#github.com/CaravanaCloud/rinha-de-backend-2024-q1-impl
